Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -77,43 +78,56 @@ instance : AddMonoid DyckWord where
 
 namespace DyckWord
 
-variable (p q : DyckWord)
+variable {p q : DyckWord}
 
 lemma toList_eq_nil : p.toList = [] ↔ p = 0 := by rw [DyckWord.ext_iff]; rfl
+lemma toList_ne_nil : p.toList ≠ [] ↔ p ≠ 0 := toList_eq_nil.ne
+
+@[simp]
+lemma add_eq_zero_iff : p + q = 0 ↔ p = 0 ∧ q = 0 := by
+  simp_rw [← toList_eq_nil]; apply append_eq_nil
```

Comment:
Actually can you prove `Subsingleton (AddUnits DyckWord)`? Then this lemma is just `add_eq_zero`

Reply:
`add_eq_zero` requires `AddCommMonoid`, and Dyck word addition is not commutative.

Reply:
I'm sure `add_eq_zero` can be generalise to `AddMonoid`. Can you do that?

Reply:
That, unfortunately, **cannot be done** – at least left-cancellativity is required (which Dyck words do satisfy). [This MSE answer](https://math.stackexchange.com/a/4535529/357390) gives a counterexample to `add_ne_zero` on arbitrary monoids as the free one on two generators with `ab=1`.

Reply:
Mind writing the cancellative version then?

Reply:
All done (including your comment below on `head_eq_U`/`getLast_eq_D`).